### ISSUE TITLE
Can't Uninstall, Context menu link remains.

### ISSUE THREAD

**Issue by isaacchristie** [falahati/NetworkAdapterSelector#11]
After testing the software I tried uninstalling it, but the installer returned the error: 
`Error 1721. There is a problem with this Windows Installer package. A program required for this install to complete could not be run. Contact your support personnel or package vendor. Action: UnregisterShellExtension, location: ..\NetworkAdap..` 
![2020-04-01_223412](https://user-images.githubusercontent.com/37149412/78150677-356e5500-746a-11ea-88aa-0581449f6314.jpg)

And the context menu reference to the program is still there
![2020-04-01_223436](https://user-images.githubusercontent.com/37149412/78150697-3bfccc80-746a-11ea-92c5-0d33211590c5.jpg)

I tried using the original installer to uninstall it and also the file from this post "https://github.com/falahati/NetworkAdapterSelector/releases/tag/v1.1.0.0-dev" both things didnt help.

**Comment by falahati** [comment]
It's unfortunate.
Did you try restarting your system? The shell extension is always open by Explorer.exe and the setup sometimes can't install or uninstall it depending on the state of the software or the Explorer. In the majority of the times a restart will help.

Some programs like [Revo Unsinstaller](https://www.revouninstaller.com/revo-uninstaller-free-download/) can forcefully remove applications. You might want to try that out while I am trying to solve this.

I am waiting on a PR on the library we use for ShellExtensions to go through from a year ago that will help a lot with this problem.

**Comment by isaacchristie** [comment]
No I had already tried restarting. Ive also tried allowing the installer to close all applications as it needed during uninstall and it did close windows explorer infact it didnt open it back up after it had completed i had to do that myself from CTRL-ALT-DEL -> Task Manager -> File -> Run New Task -> explorer.exe

I finally resorted to registry editors to clear the shell extension to get rid of the context menu item. The uninstaller is clearly broken needs to be fixed.

**Comment by realcapo** [comment]
I have the same problem. Is there an update on this? @isaacchristie what specific registry editor did you use? I tried removing it with revo but to no avail.

**Comment by XT3RM1N8R** [comment]
### **This is still very much an issue and I urge you to include a warning about this in the README and elsewhere in the meantime.** @falahati
---
_**For those of you who are having issues uninstalling this tool, please see below.**_ (@realcapo check this out)

I had to use [ShellExView](https://www.nirsoft.net/utils/shexview.html) to disable and delete the Shell Extension, including the associated registry keys. 
Look in the `Filename` column for a path that matches the `Network Adapter Selector` install location. 
Sort by the `File Created Time` column to save you some time.
The `Extension Name` column may say something like **"Dynamic Menu Item"**, but the `Filename` column will be your best bet.
When you find the extension, right-click it to `Disable Selected Item` and double-check that the context menu shell extension disappears _**after restarting Explorer**_.
If you think you found it, you can right click on the extension within ShellExView to `Open CLISD in RegEdit` and then delete the matching key(s).

I then had to delete the `Network Adapter Selector` installation within my `C:\Program Files (x86)\` folder.

_Next, you're going to want to use the `Uninstall` feature on the **"Network Adapter Selector"** entry in the Windows **Uninstall or change a program** tool within **Control Panel**. 
This feature will fail, but we can get the filepaths that appear in the **User Account Control** pop-up window. 
Just take note of the filepath and close the window or select **"No"**._
Using that, I deleted these two things:
- The "hash-coded" `MSI` file in my `C:\Windows\Installer\` folder. _Get the filepath from the pop-up._
- The "hash-coded" folder containing the `Network Adapter Selector.msi` file in my `\AppData\Local\Downloaded Installations\` folder. _Get the filepath from the pop-up. It will be different this time._

Next, I deleted the "hash-coded" registry key "folder" within `HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\`.
To find the right key, right-click & export the `\Uninstall\` "folder" somewhere, then search the `REG` file for `Network Adapater Selector` using a text editor/viewer.
When you find that string, copy the "hash-coded" ID on the line above and search the `\Uninstall\` "folder" in the registry for the ID.
When you find the key, delete it and the sub-keys.

---

This all took way more time than it should have to figure out what I needed to remove--as well as where to search and then ultimately find & delete the right things.
Had I known it would be this much trouble, I would not have used Network Adapter Selector in its current state.

@falahati I appreciate you providing this tool, yet this Uninstallation problem is something that I wish I had been aware of or at least had the intuitive foresight to check the Issues for.

**Comment by kimpurcell** [comment]
Bad uninstall procedure. Frustrating.

**Comment by sperglord8008s** [comment]
still broken...what a PITA
